prow/plugins/reward-owners/reward-owners.go

```diff
@@ -0,0 +1,115 @@
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/plugins/ownersconfig"
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	// PluginName defines this plugin's registered name.
+	PluginName    = "reward-owners"
+	RewardMessage = `
+Thanks to %s for serving the community in this new capacity!
+
+Next steps:
+1- Reach out in [slack.k8s.io](http://slack.k8s.io/) -> #sig-contribex for your special badge swag!
+
+2- Join [slack.k8s.io](http://slack.k8s.io/) -> #kubernetes-contributors in slack and [dev@kubernetes.io](https://groups.google.com/a/kubernetes.io/g/dev) for all upstream info
+
+3- Review the [community-membership.md](https://github.com/kubernetes/community/blob/master/community-membership.md) doc for your role. If for some reason you can't perform the duties associated, Emeritus is a great way to take a break! [OWNERs](https://github.com/kubernetes/community/blob/master/contributors/guide/owners.md) is another great resource for how this works.
+
+4- Look over our [governance](https://github.com/kubernetes/community/blob/master/governance.md) docs now that you are actively involved in the maintenance of the project.
+`
+)
+
+func init() {
+	plugins.RegisterPullRequestHandler(PluginName, handlePullRequest, helpProvider)
+}
+
+func helpProvider(_ *plugins.Configuration, _ []config.OrgRepo) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: fmt.Sprintf("The reward-owners plugin watches in %s and %s files for modifications and welcomes new approvers and reviewers.", ownersconfig.DefaultOwnersFile, ownersconfig.DefaultOwnersAliasesFile),
+	}
+	return pluginHelp, nil
+}
+
+type githubClient interface {
+	CreateComment(owner, repo string, number int, comment string) error
+}
+
+type ownersClient interface {
+	LoadRepoOwnersSha(org, repo, base, sha string) (repoowners.RepoOwner, error)
+}
+
+type info struct {
+	base         github.PullRequestBranch
+	head         github.PullRequestBranch
+	number       int
+	org          string
+	repo         string
+	repoFullName string
+}
+
+func handlePullRequest(pc plugins.Agent, pre github.PullRequestEvent) error {
+	if pre.Action != github.PullRequestActionOpened && pre.Action != github.PullRequestActionReopened && pre.Action != github.PullRequestActionSynchronize {
+		return nil
+	}
+
+	prInfo := info{
+		base:         pre.PullRequest.Base,
+		head:         pre.PullRequest.Head,
+		number:       pre.Number,
+		org:          pre.Repo.Owner.Login,
+		repo:         pre.Repo.Name,
+		repoFullName: pre.Repo.FullName,
+	}
+	return handle(pc.GitHubClient, pc.OwnersClient, pc.Logger, prInfo)
+}
+
+func handle(ghc githubClient, oc ownersClient, log *logrus.Entry, info info) error {
+	log.Debug("Resolving repository owners for base branch...")
+	baseRepo, err := oc.LoadRepoOwnersSha(info.org, info.repo, info.base.Ref, info.base.SHA)
```

Comment:
It may be worth it to look up the files changed by the PR via the GH API before loading the owners files. Most PRs won't change OWNERS files at all and for those PRs it would probably be preferable to risk consuming some GH API rate limit in exchange for avoiding multiple repo fetches. We do this in `verify-owners` for example.